src/quadrature/quadrature_nodal_2D.C

```diff
@@ -145,6 +145,24 @@ void QNodal::init_2D(const ElemType, unsigned int)
         return;
       }
 
+    case TRI7:
+      {
+        // We can't exactly integrate cubics with only seven nodes,
+        // but with w_i = integral(phi_i) for Lagrange shape functions
+        // phi_i, we'll at least get exact integrals of every Lagrange
+        // shape function, including the cubic bubble.
```

Comment:
I agree with the weights you found here, but I don't see how you get them by integrating the TRI7 basis functions? If I do that using sympy, I get the following:
```
import sympy as sy
from fractions import Fraction

xi, eta = sy.sympify('xi, eta')

zta0 = 1 - xi - eta
zta1 = xi
zta2 = eta

# Representative vertex basis function
phi0 = 2 * zta0 * (zta0 - Fraction(1,2)) + Fraction(3,27) * zta0 * zta1 * zta2
wv = sy.integrate(sy.integrate(phi0, (eta, 0, 1-xi)), (xi, 0, 1))
print(f'wv={wv}') # wv=1/1080

# Representative edge basis function
phi3 = 4 * zta0 * zta1 - Fraction(12,27) * zta0 * zta1 * zta2
we = sy.integrate(sy.integrate(phi3, (eta, 0, 1-xi)), (xi, 0, 1))
print(f'we={we}') # we=22/135

# Centroid basis function
phi6 = zta0 * zta1 * zta2
wc = sy.integrate(sy.integrate(phi6, (eta, 0, 1-xi)), (xi, 0, 1))
print(f'wc={wc}') # wc=1/120

# Weights should sum up to 1/2, the area of the reference element
print(f'Weight sum = {3*wv + 3*we + wc}') # weight sum = 1/2
```
Also, I believe that the rule you have here *is* exact for cubics.

Reply:
Oops, I was off by a factor of 1/27 in my original script, the corrected version agrees with you now:
```
import sympy as sy
from fractions import Fraction

xi, eta = sy.sympify('xi, eta')

zta0 = 1 - xi - eta
zta1 = xi
zta2 = eta

# Representative vertex basis function
phi0 = 2 * zta0 * (zta0 - Fraction(1,2)) + 3 * zta0 * zta1 * zta2
wv = sy.integrate(sy.integrate(phi0, (eta, 0, 1-xi)), (xi, 0, 1))
print(f'wv={wv}') # wv=1/40

# Representative edge basis function
phi3 = 4 * zta0 * zta1 - 12 * zta0 * zta1 * zta2
we = sy.integrate(sy.integrate(phi3, (eta, 0, 1-xi)), (xi, 0, 1))
print(f'we={we}') # we=1/15

# Centroid basis function
phi6 = 27 * zta0 * zta1 * zta2
wc = sy.integrate(sy.integrate(phi6, (eta, 0, 1-xi)), (xi, 0, 1))
print(f'wc={wc}') # wc=9/40
```

So I guess the only further clarification I would make is that this rule is actually exact for cubics... I'm 99% sure about this, but it also wouldn't hurt my feelings if you wanted to verify that :)

Reply:
> So I guess the only further clarification I would make is that this rule is actually exact for cubics... I'm 99% sure about this, but it also wouldn't hurt my feelings if you wanted to verify that :)

Whoa, seriously?  That's fantastic.  I didn't even bother to check for that because it seemed too unlikely to bother.

Reply:
It totally works.  Quadrature rules are magic.

I've updated the comment accordingly.

I can't actually think of any use for this fact, though.  We generally only do nodal quadrature rules when we want a lumped mass matrix, but we only get a lumped mass matrix if we're using a Lagrange basis on that exact set of nodes, so if we try to "cheat" a Tri6 by using a Tri7 rule, we wouldn't get a lumped mass matrix with one extra entry, we'd just get a dense mass matrix that only integrates exact cubics at a cost which could have given us exact quintics.

Reply:
Agreed, it's kind of cool that it works, but I'm not sure when it would ever be useful. The only potential benefit that comes to mind is in some compressible flow formulations, where you might use a nodal quadrature to evaluate inviscid flux terms for stability reasons... though such formulations are never based on TRI7 elements. :man_shrugging: